pkg/agent/nodeportlocal/portcache/port_table.go

```diff
@@ -85,25 +89,47 @@ func (pt *PortTable) GetDataForPodIP(ip string) []NodePortData {
 	return allData
 }
 
-func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) GetEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
-	return pt.getEntryByPodIPPort(ip, port)
+	return pt.getEntryByPodIPPortProtocol(ip, port, protocol)
 }
 
-func (pt *PortTable) getEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) getEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	for _, data := range pt.Table {
-		if data.PodIP == ip && data.PodPort == port {
+		if data.PodIP == ip && data.PodPort == port && data.Protocol == protocol {
 			return &data
 		}
 	}
 	return nil
 }
 
-func (pt *PortTable) getFreePort(podIP string, podPort int) (int, Closeable, error) {
+func (pt *PortTable) isNodePortClaimedForPodIPProtocol(ip string, nodeport int, protocol string) bool {
+	var val NodePortData
+	ok := false
+	// Since NPL supports only TCP and UDP protocols, we can check if an entry corresponding to key `<nodeport>:udp`
+	// is present in the PortTable, if the protocol we want to serve is TCP, and vice versa.
+	switch protocol {
+	case strings.ToLower(string(corev1.ProtocolTCP)):
+		val, ok = pt.Table[util.BuildPortProto(fmt.Sprint(nodeport), strings.ToLower(string(corev1.ProtocolUDP)))]
+	case strings.ToLower(string(corev1.ProtocolUDP)):
+		val, ok = pt.Table[util.BuildPortProto(fmt.Sprint(nodeport), strings.ToLower(string(corev1.ProtocolTCP)))]
+	}
+	if ok && val.PodIP != ip {
+		return false
+	}
+	return true
+}
+
+func (pt *PortTable) getFreePort(podIP string, podPort int, protocol string) (int, Closeable, error) {
 	for i := pt.StartPort; i <= pt.EndPort; i++ {
-		if _, ok := pt.Table[i]; !ok {
-			socket, err := pt.LocalPortOpener.OpenLocalPort(i)
+		if _, ok := pt.Table[util.BuildPortProto(fmt.Sprint(i), protocol)]; !ok {
+			// This ensures that the NodePort `i` is not taken by any other Pods,
+			// irrespective of the protocol.
+			if !pt.isNodePortClaimedForPodIPProtocol(podIP, i, protocol) {
+				continue
+			}
```

Comment:
this logic is too convoluted IMO.

The comment makes sense "not taken by any other Pods, irrespective of the protocol", but in a way it doesn't align with the fact that `isNodePortClaimedForPodIPProtocol` takes the protocol as a parameter.

I also don't think you should be reusing the same NodePort for a given Pod unless the podPort is also the same, which doesn't seem to be the case here.

I don't understand why we don't stick to the earlier logic: a (podIP, portPort) tuple maps to a unique NodePort, irrespective how which protocols are in use. There is no need to include the protocol as a key to `pt.Table`. All you need is to modify `NodePortData` to accommodate a list of protocols.

Reply:
Having the Protocol in the key helps in checking whether the nodePort can be used for the pod or not. 
```
if _, ok := pt.Table[util.BuildPortProto(fmt.Sprint(i), protocol)]; !ok {
```
If I remove the protocol, I'll have to remove the key check altogether since a single nodeport can host multiple protocols. We'll be looping over and checking nodeports from `StartPort` every time. In that sense having nodeport and protocol in the key ensures 1 iptable rule per NodePortData key, just like it was prior to the multi protocol support.
Also, initially I did not want to touch the `NodePortData` and `PodNodePort` structs and just wanted to work over the annotation structs, for minimal changes.

> I also don't think you should be reusing the same NodePort for a given Pod unless the podPort is also the same, which doesn't seem to be the case here.


I can probably include this in the checks that are being done in `pt.isNodePortClaimedForPodIPProtocol`

Reply:
I am indeed suggesting changing `NodePortData` as I think it will make the code simpler.
From:
```go
type NodePortData struct {
	NodePort int
	PodPort  int
	PodIP    string
	Protocol string
	socket   Closeable
}
```
To something like:
```go
type NodePortProtocolData struct {
        Protocol string
        socket Closeable
}

type NodePortData struct {
	NodePort int
	PodPort  int
	PodIP    string
        Protocols []NodePortProtocolData
}
```

This way the code logic stays simple and very close to what it is today, when it comes to allocating Node port mappings. It's also a simple way to show that 1 (Pod IP, Pod port) tuple maps to 1 Node port, irrespective of which protocols are "enabled" for that port.

Reply:
this has been addressed. resolving this.